### PR TITLE
#802 - Rendering annotations pagewise

### PR DESCRIPTION
**Pull request by wpmp**
**What's in the PR**
* add pagewise annotation rendering

**How to test manually**
* open a PDF document in PDF editor
* scroll to pages and see rendering

**Automatic testing**
* [ ] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

**Comment by wpmp** [comment]
PR is based on #1049 
Will also require #1058 for further testing

**Comment by wpmp** [comment]
@reckart somehow I cannot change the labels to set it to `needs review`... any clue why this is the case?

**Comment by reckart** [comment]
@wpmp I don't know. You have write permissions to the repo, so as far as I understand [the GitHub documentation](https://help.github.com/en/articles/repository-permission-levels-for-an-organization), you should be able to change issues including their labels.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java [hunk 1]
```diff
@@ -105,7 +105,8 @@ public void respond(AjaxRequestTarget aTarget)
             @Override
             protected void respond(AjaxRequestTarget aTarget)
             {
-                aPdfAnnotationEditor.renderPdfAnnoModel(aTarget);
+//                aPdfAnnotationEditor.renderPdfAnnoModel(aTarget);
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java]: WUT? Why is the above line commented out and a message logged to system out? If you want to log a warning about an unexpected API call, better log it through the logging system. Or throw an exception... or (since you are in a AJAX call) you could even log a feedback message to the UI.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Offset.java [hunk 2]
```diff
@@ -72,4 +72,9 @@ public String toString() {
             ", end=" + end +
             '}';
     }
+
+    public static Offset empty()
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/Offset.java]: How about using singleton here?

```
static final Offset EMPTY = new Offset(...);

Offset empty() {
  return EMPTY;
}
```

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java [hunk 3]
```diff
@@ -199,6 +209,12 @@ private void initializeStringContent(String aPdftxt)
...
 
+            if (page > lastPage) {
+                pageOffsetMap.put(lastPage, new Offset(pageBeginIndex, extractLineIndex - 1));
+                lastPage = page;
+                pageBeginIndex = extractLineIndex;
+            }
+
             // if value of PdfExtractLine is in brackets it is a draw operation and is ignored
             if (!extractLine.getValue().matches("^\\[.*\\]$")
                 && !extractLine.getValue().equals("NO_UNICODE"))
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: Can you add a comment when `NO_UNICODE` appears?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java [hunk 4]
```diff
@@ -260,4 +279,19 @@ public Offset getStringIndex(int aExtractIndex)
             return new Offset(sanitizedIndex, sanitizedIndex);
         }
     }
+
+    /**
+     * Get gegin and end offset of a page.
+     * @param page
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: Either document parameter and return value with a proper description or remove the param/return stanzas.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/render/PdfAnnoRenderer.java [hunk 5]
```diff
@@ -187,8 +187,7 @@ private static String getColor(VObject aVObject, ColoringStrategy aColoringStrat
     /**
      * for each span adds context with given windowsize before and after span.
      */
-    private static void addContextToSpans(List<RenderSpan> aSpans, int windowSize,
-                                          String text)
+    private static void addContextToSpans(List<RenderSpan> aSpans, int windowSize, String text)
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/render/PdfAnnoRenderer.java]: `aWindowSize` + `aText`.

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js [hunk 6]
```diff
@@ -108,6 +108,24 @@ window.addEventListener('DOMContentLoaded', async () => {
...
+    "action": "getAnnotations",
+    "page": window.pageNumber
+  }
+  parent.Wicket.Ajax.ajax({
+    "m": "POST",
+    "ep": data,
+    "u": window.apiUrl,
+    "sh": [],
+    "fh": [function () {
+      console.log('Something went wrong on requesting annotations from inception backend.')
```
Comment (reckart) [inline inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: Wouldn't it be good to show this to the user, e.g. via a JS popup?

File: inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java [hunk 7]
```diff
@@ -171,7 +171,7 @@ public void testRender() throws Exception
 
         PdfExtractFile pdfExtractFile = new PdfExtractFile(pdftxt, new HashMap<>());
         PdfAnnoModel annoFile = PdfAnnoRenderer.render(state, vdoc,
-            cas.getDocumentText(), schemaService, pdfExtractFile);
+            cas.getDocumentText(), schemaService, pdfExtractFile, Offset.empty());
```
Comment (reckart) [inline inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java]: If you pass `Offset.empty()` here, will the annoFile contain nothing?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/render/RenderSpan.java [hunk 8]
```diff
@@ -44,13 +44,15 @@ public RenderSpan(Offset aOffset)
         end = aOffset.getEnd();
     }
 
-    public RenderSpan(VSpan aVSpan, Span aSpan)
+    public RenderSpan(VSpan aVSpan, Span aSpan, Offset offset)
```
Comment (reckart) [inline inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/render/RenderSpan.java]: Why pass an offset with begin/end if you only ever use the begin? I'd suggest to only pass the begin to make the code easier to understand and maintain.

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java (hunk 1 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoPanel.java]: I think I need more sleep.. this shouldn't have been commited..
This whole `AbstractDefaultAjaxBehavior` should be removed as all calls regarding annotations are now done over the `apiProvider`.

File: inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java (hunk 7 above)
Comment (wpmp) [reply inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java]: No, this offset basically is the page offset.
E.g. if we render page 5 of a document this would contain the character offset to the beginning of the 5th page.
As in the test case we render the whole document we don't need an offset.
I probably should add tests for this with offset other than 0

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js [hunk 9]
```diff
@@ -189,6 +211,25 @@ async function displayViewer () {
     textLayer.setup(analyzeResult)
     window.annoPage.pdftxt = analyzeResult
 
+// BEGIN INCEpTION EXTENSION - #802 - Rendering annotations pagewise
+    const renderTimeout = 1500
```
Comment (wpmp) [inline inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: @reckart this is the timeout that PDFAnno will wait until it renders a new page if the page hasn't changed since last change.
Do you think 1500 ms is a good value?
I don't know if it is too short or too long.

File: inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java (hunk 7 above)
Comment (reckart) [reply inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java]: If 0 is a valid offset, then maybe `Offset.empty()` should be `[-1,-1]` ? Or what is the purpose of this empty offset? I assume it is used to indicate a default invalid value.

Comment (reckart) [reply inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java]: Some JavaDoc would be good here ;)

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js (hunk 9 above)
Comment (reckart) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: > it renders a new page if the page hasn't changed since last change

I'm not sure I understand this. What does it mean if a page is "new" or has been "changed" and what kind of change is the "last change" - probably a different kind of change than the change to the page...?

File: inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java (hunk 7 above)
Comment (wpmp) [reply inception-pdf-editor/src/test/java/de/tudarmstadt/ukpinception/pdfeditor/pdfanno/model/PdfAnnoRendererTest.java]: No it is fine as it is, it just indicates we do not have some offset for rendering here.
The offset begin value is added to each span.
If we would have like page 2, the offset would be the index of page 2 begin.
Here it is just at the begin of the document, hence offset 0.
Maybe `empty` is a misleading name.
But as you stated somewhere else in the review we don't need an offset here, just the begin of the offset is required.
I will make changes according to the review now and will add a short comment on this test.

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js (hunk 9 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: When you scroll through a PDF document the editor will not render annotations immediately for each page but it will wait for the `renderTimeout` until it renders annotations.
The reason is if you scroll through the document quickly you don't want a render request for each page, but only when you stop at some page and stay on it.
So the question is, how long should you have to wait until editor requests annotations.

Comment (reckart) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: Things appear to be "interactive" if the reaction time is 200ms or less. In some places in our code, we use a delay of 500ms. Using a longer delay may be sensible. By which rationale did you choose 1500ms?

Comment (wpmp) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: I chose it like this because scrolling through a page takes some time, so if it is too short it might trigger rendering too often.
That was my thought.
So when it renders the page will hang and the user can't continue scrolling through the document.
I think this could be annoying.

Comment (reckart) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: Is there a way to render n page ahead and start requesting new ones when the user is e.g. half-way through the already rendered pages?

Comment (wpmp) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: Still it would hang though.
Also the rendering process would take longer I guess.
Currently previous and next page is rendered along to the current page so that you can see annotations at page boundaries.

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js (hunk 6 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: With `window.alert()`?
Or do we have Wicket, Webanno or some other stuff for this?

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js (hunk 9 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: I think I found a cooler way.
What about if we render only when page changes and no movement in document is detected for 500ms ?
It feels pretty smooth to me and you don't have to scroll too fast through document if you want to get to a certain point without interruption.
I will commit this, then you can test and tell me then whether you like it or not, ok?

File: inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java (hunk 3 above)
Comment (wpmp) [reply inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/model/PdfExtractFile.java]: It is in the PdfExtract file.
It is used "when unicode mapping is unavailable".

**Comment by wpmp** [comment]
Ah, changing labels now works again. :)

File: inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js (hunk 9 above)
Comment (reckart) [reply inception-pdf-editor/src/main/js/pdfanno/src/pdfanno.js]: It sounds very good. Looking forward to testing it.

**Comment by wpmp** [comment]
@reckart seems like the tests broke with webanno update?

**Comment by wpmp** [comment]
I think it is fixed with removal of `state.setFirstVisibleUnit` usage.

**Comment by reckart** [comment]
My guess would be that you should call `state.setPagingStrategy(new SentenceOrientedPagingStrategy())` whenever you create an `AnnotatorStateImpl` to properly fix this.

**Comment by wpmp** [comment]
Jenkins, can you test this please?

**Comment by wpmp** [comment]
> My guess would be that you should call `state.setPagingStrategy(new SentenceOrientedPagingStrategy())` whenever you create an `AnnotatorStateImpl` to properly fix this.

Changed it like this.

**Comment by reckart** [comment]
Nice 👍
